RecoLocalCalo/EcalRecProducers/plugins/EcalPhase2DigiToGPUProducer.cc

```diff
@@ -0,0 +1,103 @@
+
+  cms::cuda::ContextState cudaState_;
+};
+
+void EcalPhase2DigiToGPUProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("BarrelDigis", edm::InputTag("simEcalUnsuppressedDigis", ""));
+  desc.add<std::string>("digisLabelEB", "ebDigis");
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+EcalPhase2DigiToGPUProducer::EcalPhase2DigiToGPUProducer(const edm::ParameterSet& ps)
+    : digiCollectionToken_(consumes<EBDigiCollectionPh2>(ps.getParameter<edm::InputTag>("BarrelDigis"))),
+      digisCollectionToken_(produces<cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>>(
+          ps.getParameter<std::string>("digisLabelEB"))) {}
+
+void EcalPhase2DigiToGPUProducer::acquire(edm::Event const& event,
+                                          edm::EventSetup const& setup,
+                                          edm::WaitingTaskWithArenaHolder holder) {
+  cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(holder), cudaState_};
+
+  //input data from event
+  const auto& pdigis = event.get(digiCollectionToken_);
+
+  size_ = pdigis.size();
+
+  //allocate device pointers for output
+  digis_.ids = cms::cuda::make_device_unique<uint32_t[]>(size_, ctx.stream());
+  digis_.data = cms::cuda::make_device_unique<uint16_t[]>(size_ * EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
+
+  //allocate host pointers for holding product data and id vectors
+  auto idstmp = cms::cuda::make_host_unique<uint32_t[]>(size_, ctx.stream());
+  auto datatmp = cms::cuda::make_host_unique<uint16_t[]>(size_ * EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
+
+  //iterate over digis
+  uint32_t i = 0;
+  for (const auto& pdigi : pdigis) {
+    const int nSamples = pdigi.size();
+    //assign id to output vector
+    idstmp.get()[i] = pdigi.id();
+    //iterate over sample in digi
+    for (int sample = 0; sample < nSamples; ++sample) {
+      //get samples from input digi
+      EcalLiteDTUSample thisSample = pdigi[sample];
+      //assign adc data to output
+      datatmp.get()[i * nSamples + sample] = thisSample.raw();
+    }
+    ++i;
+  }
+
+  //copy output vectors into member variable device pointers for the output struct
+
+  cudaCheck(
+      cudaMemcpyAsync(digis_.ids.get(), idstmp.get(), size_ * sizeof(uint32_t), cudaMemcpyHostToDevice, ctx.stream()));
+  cudaCheck(cudaMemcpyAsync(digis_.data.get(),
+                            datatmp.get(),
+                            size_ * EcalDataFrame_Ph2::MAXSAMPLES * sizeof(uint16_t),
+                            cudaMemcpyHostToDevice,
+                            ctx.stream()));
+}
+
+void EcalPhase2DigiToGPUProducer::produce(edm::Event& event, edm::EventSetup const& setup) {
+  //get cuda context state for producer
+  cms::cuda::ScopedContextProduce ctx{cudaState_};
+  digis_.size = size_;
+
+  //emplace output in the context
+  ctx.emplace(event, digisCollectionToken_, std::move(digis_));
```

Comment:
In fact, I _think_ this could also be done asynchronously, without the `acquire()`/`produce()` split.

@makortel is the expert here, but could you try to make the change and see if it works ?
  - change the inheritance to simply `edm::stream::EDProducer<>`
  - move the code from `acquire()` to the beginning of `produce()`
  - change
    ```diff
    -cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(holder), cudaState_};
    +cms::cuda::ScopedContextProduce ctx{event.streamID()};
    ```
  - make `size_` and `digis_` local variables 
  - remove the `cudaState_` variable

?

Reply:
The possibility of dropping the `ExternalWork` was something that has been mentioned already. However, since `holder` was used to construct `ctx` we thought that it does not work in the end since `holder` is not available in `produce()`. But using `ctx` with type `cms::cuda::ScopedContextProduce` could make that possible.

Reply:
> In fact, I _think_ this could also be done asynchronously, without the `acquire()`/`produce()` split.
> 
> @makortel is the expert here, but could you try to make the change and see if it works ?

For what it's worth, I agree.

> But using `ctx` with type `cms::cuda::ScopedContextProduce` could make that possible.

Right, without `ExternalWork` the `ScopedContextProduce` should be used.